### PR TITLE
Replace gRPC code generation tool from Znly/protoc to Buf

### LINKED ISSUE #2141: Replace the tool of code generator for the gRPC API
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
Currently, we are generating codes for the gRPC API from the protocol buffers definition with [`znly/protoc`](https://hub.docker.com/r/znly/protoc/).

However, the image doesn't seem to be maintained since the latest image was published 4 years ago.

So, we should replace the toll with another one.

follow-up: #2140 

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

---

Love this feature? Give it a 👍 We prioritize the features with the most 👍

### PR DESCRIPTION
**Pull request by shivas1516**
This commit updates the codebase to utilize Buf for generating gRPC code instead of Zync.

**What this PR does / why we need it**:

This PR replaces the outdated code generation tool 'znly/protoc' with Buf for generating gRPC code. The change is necessary due to the lack of maintenance and updates for 'znly/protoc', ensuring compatibility, security, and access to new features with Buf.

**Which issue(s) this PR fixes** _(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)_:
Fixes #2141

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/components/katib/) included if any changes are user facing

### CODE REVIEW

**Comment by tenzen-y** [comment]
@shivas1516 Thank you for this contribution!
First of all, could you sign cla and DCO?

**Comment by shivas1516** [comment]
Of course! I'm glad you found my contribution valuable. Before proceeding, may I ask if you could kindly provide more context or details regarding the CLA and DCO you mentioned? That way, I can better assist you with the signing process.

**Comment by tenzen-y** [comment]
> Of course! I'm glad you found my contribution valuable. Before proceeding, may I ask if you could kindly provide more context or details regarding the CLA and DCO you mentioned? That way, I can better assist you with the signing process.

Sure.

CLA singing step: https://github.com/kubeflow/katib/pull/2277/checks?check_run_id=22474756817
DCO signing step: https://github.com/kubeflow/katib/pull/2277/checks?check_run_id=22474756788

**Comment by shivas1516** [comment]
> > Of course! I'm glad you found my contribution valuable. Before proceeding, may I ask if you could kindly provide more context or details regarding the CLA and DCO you mentioned? That way, I can better assist you with the signing process.
> 
> Sure.
> 
> CLA singing step: https://github.com/kubeflow/katib/pull/2277/checks?check_run_id=22474756817 DCO signing step: https://github.com/kubeflow/katib/pull/2277/checks?check_run_id=22474756788

I really appreciate your time and patience. Just to let you know, I've signed both the CLA and the DCO. Thanks again for your help!

**Comment by tenzen-y** [comment]
@kubeflow/wg-automl-leads 
Could you approve CI? Checking the result of CI would be helpful before we review this.

**Comment by shivas1516** [comment]
Ofcourse, Let me rework on this CI errors. If any Suggestions regarding this CI Errors please let me know, That will help me to move forward in a correct way faster.

**Comment by shivas1516** [comment]
> @shivas1516 Could you address CI errors?

Ofcourse, Let me rework on this CI errors. If any Suggestions regarding this CI Errors please let me know, That will help me to move forward in a correct way faster.

File: pkg/apis/manager/v1beta1/build.sh [hunk 1]
```diff
@@ -1,29 +0,0 @@
-#!/usr/bin/env bash
```
Comment (andreyvelich) [inline pkg/apis/manager/v1beta1/build.sh:1]: How we can use this new tool to re-generate files ?

**Review by andreyvelich — COMMENTED** [review]
Thank you for your contribution @shivas1516 🎉 
Did we investigate what tools other Kubeflow components are using to generate protobufs ?
For example, Kubeflow Pipelines ?

**Comment by shivas1516** [comment]
> Thank you for your contribution @shivas1516 🎉 Did we investigate what tools other Kubeflow components are using to generate protobufs ? For example, Kubeflow Pipelines ?

@andreyvelich Unfortunately, I'm just getting back to you now. I've been tied up addressing the CI errors in this PR. However, I'm having difficulty grasping your thoughts clearly. Could you please provide some detailed explanation for better clarity?

**Comment by andreyvelich** [comment]
> > Thank you for your contribution @shivas1516 🎉 Did we investigate what tools other Kubeflow components are using to generate protobufs ? For example, Kubeflow Pipelines ?
> 
> @andreyvelich Unfortunately, I'm just getting back to you now. I've been tied up addressing the CI errors in this PR. However, I'm having difficulty grasping your thoughts clearly. Could you please provide some detailed explanation for better clarity?

@shivas1516 Currently, if we make any changes in [`api.proto`](https://github.com/kubeflow/katib/blob/master/pkg/apis/manager/v1beta1/api.proto), we can use `make generate` to re-generate files. Since you are going to remove `pkg/apis/manager/health/build.sh` file, how we can re-generate gRPC files ?

**Comment by shivas1516** [comment]
> Since

@andreyvelich When migrating to Buf, instead of using `make generate`, we must utilize the `buf generate` command to regenerate code changes in `.proto` files. This command relies on configurations specified in `buf.gen.yaml`, where we can define language preferences for generating code, remote plugins to use, and versions for our convenience.

Hope I'm clear to you, If any suggestions or need clarity regarding this. Please let me know.

**Comment by shivas1516** [comment]
> > > Thank you for your contribution @shivas1516 🎉 Did we investigate what tools other Kubeflow components are using to generate protobufs ? For example, Kubeflow Pipelines ?
> > 
> > 
> > @andreyvelich Unfortunately, I'm just getting back to you now. I've been tied up addressing the CI errors in this PR. However, I'm having difficulty grasping your thoughts clearly. Could you please provide some detailed explanation for better clarity?
> 
> @shivas1516 Currently, if we make any changes in [`api.proto`](https://github.com/kubeflow/katib/blob/master/pkg/apis/manager/v1beta1/api.proto), we can run `make generate` to re-generate files. Since you are going to remove `pkg/apis/manager/health/build.sh` file, how we can re-generate gRPC files ?


@andreyvelich When migrating to Buf, instead of using make generate, we must utilize the buf generate command to regenerate code changes in .proto files. This command relies on configurations specified in buf.gen.yaml, where we can define language preferences for generating code, remote plugins to use, and versions for our convenience.

Hope I'm clear to you, If any suggestions or need clarity regarding this. Please let me know.

File: pkg/apis/manager/v1beta1/build.sh (hunk 1 above)
Comment (shivas1516) [reply pkg/apis/manager/v1beta1/build.sh:1]: we utilize the `buf generate` command to regenerate code changes in `.proto`

**Comment by tenzen-y** [comment]
> When migrating to Buf, instead of using `make generate`, we must utilize the `buf generate` command to regenerate code changes in `.proto` files. This command relies on configurations specified in `buf.gen.yaml`, where we can define language preferences for generating code, remote plugins to use, and versions for our convenience.
> 
> Hope I'm clear to you, If any suggestions or need clarity regarding this. Please let me know.

Any reasons that you closed this PR?

**Comment by shivas1516** [comment]
> > When migrating to Buf, instead of using `make generate`, we must utilize the `buf generate` command to regenerate code changes in `.proto` files. This command relies on configurations specified in `buf.gen.yaml`, where we can define language preferences for generating code, remote plugins to use, and versions for our convenience.
> > 
> > Hope I'm clear to you, If any suggestions or need clarity regarding this. Please let me know.
> 
> @shivas1516 Any reasons that you closed this PR?
> 



Thank you for your inquiry regarding the closure of the PR.

Unfortunately, the PR was closed due to an error that occurred in the my forked Katib repository GitHub codespace. Regrettably, I had to delete the forked repository master as a result. However, I am currently reworking on this issue, and I am confident that it will be even better this time around. I have reached out to the Buf community and received valuable suggestions from them, which I believe will contribute to the improvement of the project.

**Comment by tenzen-y** [comment]
> > > When migrating to Buf, instead of using `make generate`, we must utilize the `buf generate` command to regenerate code changes in `.proto` files. This command relies on configurations specified in `buf.gen.yaml`, where we can define language preferences for generating code, remote plugins to use, and versions for our convenience.
> > > Hope I'm clear to you, If any suggestions or need clarity regarding this. Please let me know.
> > 
> > 
> > @shivas1516 Any reasons that you closed this PR?
> 
> Thank you for your inquiry regarding the closure of the PR.
> 
> Unfortunately, the PR was closed due to an error that occurred in the my forked Katib repository GitHub codespace. Regrettably, I had to delete the forked repository master as a result. However, I am currently reworking on this issue, and I am confident that it will be even better this time around. I have reached out to the Buf community and received valuable suggestions from them, which I believe will contribute to the improvement of the project.

Thank you for your effort. I just wanted to confirm if you're still in here.

**Pull request closed without merging**